src/Microsoft.DotNet.Build.Tasks.Feed/BlobFeedAction.cs

```diff
@@ -0,0 +1,179 @@
+                    }
+                    Func<string, string> calculateIndexPath = (string item) =>
+                    {
+                        return item.Substring(IndexDirectory.Length + 1).ToLowerInvariant();
+                    };
+                    await Task.WhenAll(indexList.Select(item => UploadAsync(CancellationToken, item, calculateIndexPath(item), clientThrottle, listAzureBlobs)));
+                }
+            }
+
+            return !Log.HasLoggedErrors;
+        }
+
+        public async Task<bool> PushToFeedFlat(IEnumerable<string> items, string relativePath)
+        {
+            using (var clientThrottle = new SemaphoreSlim(this.MaxClients, this.MaxClients))
+            {
+                List<string> listAzureBlobs = await ListAzureBlobs.ListBlobs(Log, feed.AccountName, feed.AccountKey, feed.ContainerName, string.Empty);
+                Func<string, string> calculatePath = (string item) =>
+                {
+                    return item.Substring(item.IndexOf(relativePath)).Replace("\\", "/");
+                };
+                await Task.WhenAll(items.Select(item => UploadAsync(CancellationToken, item, calculatePath(item), clientThrottle, listAzureBlobs)));
+            }
+            return !Log.HasLoggedErrors;
+        }
+
+        private async Task UploadAsync(CancellationToken ct, string item, string uploadPath, SemaphoreSlim clientThrottle, List<string> listAzureBlobs)
+        {
+            if (!File.Exists(item))
+                throw new Exception(string.Format("The file '{0}' does not exist.", item));
+
+            await clientThrottle.WaitAsync();
+            string leaseId = string.Empty;
+            bool isLeaseRequired = listAzureBlobs.Contains(uploadPath);
+            AzureBlobLease blobLease = new AzureBlobLease(feed.AccountName, feed.AccountKey, string.Empty, feed.ContainerName, uploadPath, Log, "60", "10");
+            if (isLeaseRequired)
+            {
+                try
+                {
+                    leaseId = blobLease.Acquire();
+                }
+                catch (Exception)
+                {
+                    Log.LogError($"Unable to obtain lease on {uploadPath}");
+                }
+            }
+
+            try
+            {
+                Log.LogMessage($"Uploading {item} to {uploadPath}.");
+                UploadClient uploadClient = new UploadClient(Log);
+                await
+                    uploadClient.UploadBlockBlobAsync(
+                        ct,
+                        feed.AccountName,
+                        feed.AccountKey,
+                        feed.ContainerName,
+                        item,
+                        uploadPath,
+                        leaseId);
+            }
+            finally
+            {
+                clientThrottle.Release();
+            }
+
+            try
+            {
+                if (isLeaseRequired)
+                {
+                    blobLease.Release();
```

Comment:
This release should be in the finally block above. We need to ensure that we release the lease if anything bad happens in the Upload as well.